### ISSUE TITLE
Improve hourglass performances

### ISSUE THREAD

**Issue by NathanMolinier** [spinalcordtoolbox/disc-labeling-hourglass#28]
## Description

This issue is meant to discuss about strategies to improve the performances of the hourglass network.

**Comment by NathanMolinier** [comment]
## Strategy 1: Improve the loss function

### Problem

Since the addition of new images (representing different field of view of the spine) performances of the hourglass have dropped significantly (mainly due to the addition of new classes ~25 discs) leading to an increasing number of false positive detections.

This problem was partially solved [here](https://github.com/spinalcordtoolbox/disc-labeling-hourglass/issues/23#issuecomment-1721368835) but still numerous false positive detection are still occurring.

### Idea

Split the loss function into two main components:
- an accuracy loss (MSE) for masks where a disc should be predicted
- a true detection loss (MSE) to ensure that non present discs will lead to an empty prediction mask

**Comment by NathanMolinier** [comment]
## Strategy 2: Use random crop with larger fields of view

### Idea 

To improve the performances and help the model to generalize, a new data augmentation function `random_crop` was added to randomly vary the discs that are present in the image.

### Results

The task still is too difficult for the model, the number of different classes is maybe to important.

**Comment by NathanMolinier** [comment]
## Strategy 3: Train 2 different model

### Idea 

Finding and identifying all the 25 discs classes seems to be to difficult for one model, therefore, I will try to train 2 models:
- one able to identify all the discs present in the image
- one able to assign classes to these detected discs

These models will be train with and without the random cropping to evaluate their ability to generalize.

### Results

...

**Comment by NathanMolinier** [comment]
## Strategy 4:

### Idea 
> based on https://www.sciencedirect.com/science/article/pii/S1361841521003030 

Retry with the 25 classes but this time with a fixed size sliding window.

The use of the Binary Cross Entropy will also be investigated to avoid having multiple class for the same disc.

### Results

...
